cartography/intel/gcp/compute.py

```diff
@@ -0,0 +1,118 @@
+    """
+    try:
+        req = compute.zones().list(project=project_id, maxResults=max_results)
+        res = req.execute()
+        return res['items']
+    except HttpError as e:
+        reason = json.loads(e.content)['error']['errors'][0]['reason']
+        if reason == 'accessNotConfigured':
+            logger.debug(
+                (
+                    "Google Compute Engine API access is not configured for project %s. "
+                    "Full details: %s"
+                ),
+                project_id,
+                e
+            )
+            return None
+        elif reason == 'notFound':
+            logger.debug(
+                (
+                    "Project %s returned a 404 not found error. "
+                    "Full details: %s"
+                ),
+                project_id,
+                e
+            )
+            return None
+        else:
+            logger.error("Could not use Compute Engine API on project %s; Reason: %s".format(project_id, reason))
+            raise e
+
+
+def get_gcp_instances_in_project(project_id, compute):
+    """
+    Return list of all GCP instances in a given project regardless of zone.
+    :param project_id: The project id
+    :param compute: The compute resource object created by googleapiclient.discovery.build()
+    :return: List of all GCP instances in given project regardless of zone.
+    """
+    zones = get_zones_in_project(project_id, compute)
+    if not zones:
+        # If the Compute Engine API is not enabled for a project, there are no zones and therefore no instances.
+        return []
+    instances = []
+    for zone in zones:
+        req = compute.instances().list(project=project_id, zone=zone['name'])
+        res = req.execute()
+        zone_instances = res.get('items', [])
+        for instance in zone_instances:
+            # Include project_id and zone_name on the `instance` dict so we can load it more easily
+            instance['project_id'] = project_id
+            instance['zone_name'] = zone['name']
+            instances.append(instance)
+    return instances
+
+
+def load_gcp_instances(neo4j_session, data, gcp_update_tag):
+    query = """
+    MATCH (p:GCPProject{id:{ProjectId}})
+    MERGE (i:GCPInstance{id:{InstanceId}})
+    ON CREATE SET i.firstseen = timestamp()
+    SET i.displayname = {DisplayName},
+    i.hostname = {Hostname},
+    i.zone_name = {ZoneName},
+    i.lastupdated = {gcp_update_tag}
+    WITH i, p
+    MERGE (p)-[r:RESOURCE]->(i)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {gcp_update_tag}
+    """
+    for instance in data:
+        neo4j_session.run(
+            query,
+            ProjectId=instance['project_id'],
+            InstanceId=instance['id'],
+            DisplayName=instance.get('name', None),
+            Hostname=instance.get('hostname', None),
+            ZoneName=instance['zone_name'],
+            gcp_update_tag=gcp_update_tag
+        )
```

Comment:
Non-blocking: are there any pieces of data that apply to compute instances across cloud providers? Would be cool if we could define an abstract `:Instance` label with a limited but cross-cloud data set.